text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+guarantee and each operation, for a total of 10 marker traits: 2 guarantees x 4 groups of methods + 2 guarantees on
+moves. Such a fine-grained approach is used in C++, and I remember writing generic methods which would static-assert
+that the elements they manipulate need be noexcept-movable, noexcept-move-assignable, and noexcept-destructible, then
+further divide the method based on whether the elements were noexcept-copyable and trivially copyable. There always is
+the nagging doubt of having missed one key guarantee, and therefore while conductive to writing finely tuned code, it
+is unfortunately not conductive to writing robust code: the risk of error is too high.
+
+The current set of traits has thus been conceived to provide a reasonable trade-off:
+
+-   A small enough number of markers that developers of collections are not overwhelmed, and thus less likely to miss
+    a key requirement leading to unsoundness in their unsafe code.
+-   A split on "natural" boundaries: one hierarchy for handle invalidation and one hierarchy for pointer invalidation.
+
+From there, feedback can be gathered as to whether further splitting or merging should be considered before
+stabilization.
+
+
+#   Prior Art
+
+##  C++
+
+In C++, `std::allocator_traits` allows one to create an Allocator which uses handles that are not (strictly) pointers.
+
+The impetus behind this design was to allow greater flexibility, much like this proposal, unfortunately it failed
+spectacularly:
+
+1.  While one can specify a non-pointer `pointer` type, this type MUST still be pointer-like: it must be
+    dereferenceable, etc... This requirement mostly requires the type to embed a pointer -- possibly augmented -- and
+    thus makes it unsuitable for in-line store, unsuitable for compaction, and only usable for shared memory usage
+    with global/thread-local companion state.
+2.  While one can specify a non-reference `reference` type, the lack of `Deref` means that such a type does not,
+    actually, behave as a reference, and while proxies can be crafted for specific types (`std::vector<bool>`
+    demonstrates it) it's impossible to craft transparent proxies in the generic case.
+
+The mistake made in the C++ allocator API is to require returning pointer-like/reference-like objects directly usable
+by the user of the collection based upon the allocator.
+
+This RFC learns from C++ mistake by introducing a level of indirection:
+
+1.  An opaque `Handle` is returned by the `Store`, which can be stored and copied freely, but cannot be dereferenced.
+    It is intended to be kept as an implementation detail within the collection, and invisible to the final user.
+2.  A `resolve` method to resolve a `Handle` into a pointer, and from there into a reference.
+
+Throwing in flexible invalidation guarantees ties the knot, allowing this API to be much more flexible than the C++
+allocator API.
```

Comment:
I must admit having always considered such shenanigans more annoying than useful, since as a user it then restricts what I can do with the type, and I typically know my usecase best ;)

Still, I'll take any suggestion to tone down/specify the phrasing to make it clear that I specifically reference the lack of flexibility of `std::allocator_traits<>::pointer` and not necessarily `std::allocator` as a whole.